### ISSUE TITLE
Using Moleculer Apollo Server in a NATS configured environment with multiple replicas.

### ISSUE THREAD

**Issue by abdavid** [moleculerjs/moleculer-apollo-server#83]
I stumbled upon this issue today and I did a bit of debugging. 

Ill try to simplify the setup for you. I have three nodes where two runs the apollo server and the last one just regular moleculerjs. 

With one instance of apollo-server and moleculerjs everything runs just fine, but when I introduce another apollo-server instance in to the pool the last one added wont start up and starts throwing errors about not being able to compile the graphl schema -- which is at first a bit weird since one instance is already running fine and doing its job. So after a bit of debugging I have over this line right here: 

https://github.com/moleculerjs/moleculer-apollo-server/blob/master/src/service.js#L377

And from what I can see this means that `generateGraphQLSchema` would actually try to generate a schema based on the already running instance of apollo-server. Is that by design?

I tried adding modifying the guard to be:

`if (processedServices.has(serviceName) || !service.local) return;` so we only generate schemas based on local services, and that works for me. 

Bug or did I do something weird here?

**Comment by abdavid** [comment]
Sorry for the delay in response here -- times are hectic.

@shawnmcknight 
While true I think it best if I draw you a diagram, and you might also be right about something not being used as intended since the moleculer-server-apollo almost always starts with throwing errors about not being able to build the schema initially, and after a few seconds (looks like service dependencies are avaliable) it recompiles and everything is fine. 

[Simplified diagram](https://viewer.diagrams.net/?highlight=0000ff&edit=_blank&layers=1&nav=1&title=Untitled%20Diagram.drawio#R5VddU5wwFP01%2B6gDCSz00V21nelY2%2FGh%2BhghQmwgNARh%2B%2Bt7syR8O2vVqU47s8OSk9wk95yTD1Z4mzUfJSnSCxFTvkJO3Kzw6Qohd40c%2BNPIziCO57dIIllssB64Yr%2BobWjQisW0HDVUQnDFijEYiTynkRphREpRj5vdCT4etSAJnQFXEeFz9DuLVdqinot7%2FBNlSWpG9kNTkRHb1iRSpiQW9QDCZyu8lUKo9i1rtpRr8iwtbdz5I7XdvCTN1VMCvqzx%2Fc%2Fym4yK7Ly5vD75fH55f2R6eSC8Mvle0LLUlCBnI8UPKs3k1c4SAr0C91DY1ClT9Kogka6pQX%2FAUpVxKLnwSsqiFeSONRQmsTFjUalo82gSbkcNeIqKjCq5gyadoYI2xNgJW5vUvTahgdKBLBYjxg1J13PPGLwY0v6AQDQnUHAaVZzKo5rertCawxQ2t8DjOlF7EqZIH0AKwbk4Kql80MQfDl0h%2BDmaA4cL8KxWCoJZBCtmqtsbqIXGYnlzrVwfz8XqwFdXC79crZZyvdacorrlLHpvnGPnMOkLK6QDX510f0b6fE%2FJ4xO9V0Mp4qQsgVW8KRWRag4P%2BANK5O4aCs6xb4s3w7pTTYrTlXa21DB1bdvB%2B80A70N0wUa0M6bx7KyYyAJZiUpG9PCWC8klVB3y6lzmgY5LMlpMUk4UexhPd0laM8JXwSCR3kXh2EUomLijTdNEDU%2BdSUd4YsfuOLQdtTzMOto7rUv7%2BeZbv1vzOe%2FZfOhNzef5xyj0UOAiL%2FDWyN617JmP3ePA9QMUBj7ycRAGz7Sm64%2Bt6fxdawb%2FwWHkjVe%2F%2F%2BaHUTgjvdkbtgDyyJwzSF2NaSqVvhZvBRcSkFzk%2BiZ8xzifQISzJNd7B1AFlzm80UTCGPzEVGQsjvlj12gpqjzWKgw2gJcpEY7NvnBv9haEmO7Wr6bDh5kObq%2BD%2FpyrpE6Nw0hbPc2U5loZ4I%2Flyf4bUD9q%2FWRlWWm7OyKCqH9XwslJuqDg4lp6hoRQ7L9L2%2B2u%2F7rHZ78B)

While developing these new services I have chosen to use the monolithic approach to start off with since this requires less configuration when deploying to k8s. What I want to achieve is to have two replicas of the moleculer-apollo-server so I can scale the monolith horizontally. What I experience when I try to do that locally is that the moleculer-apollo-server that is already running has no problems, but the newly created instance will fail when trying to generate the schema due to it identifying non-local (to that instance) services that includes the graphql configuration and thus when parsing that configuration errors will get thrown due to the fact that the same service already has been processed and duplicate types then occurs. The hack I did basically just ensures that the moleculer-apollo-server will not parse external services, only local services. Is this issue due to some services being available through the transporter?

**Comment by shawnmcknight** [comment]
So I think we can start with getting one thing out of the way:
>The hack I did basically just ensures that the moleculer-apollo-server will not parse external services, only local services.

This would be counter to the goals of `moleculer-apollo-server` in that each remote service should have its schema stitched into one composite GraphQL schema available at the gateway.  Therefore, we can't possibly skip remote services as those services would then become inaccessible in the schema.

In looking through the workflow of the schema generation, its not really clear to me how a problem would be occurring.  The new replica should start up, and each service should be discovered (local and remote).  In [generateGraphQLSchema](https://github.com/moleculerjs/moleculer-apollo-server/blob/298428592e4da9f1e3b418b6b4f9832a71598467/src/service.js#L344), each service is processed one at a time, building the schema.  When a service gets processed, it gets added to a `Set` with the service's full name (version + service name) the key within the `Set` [here](https://github.com/moleculerjs/moleculer-apollo-server/blob/298428592e4da9f1e3b418b6b4f9832a71598467/src/service.js#L377).  If the full service name is encountered any additional times, it should be skipped and nothing is added to the generated schema.  Since the service's identifier doesn't take which node its running on into account, it should be irrelevant whether that service is running locally or remotely to the `moleculer-apollo-server` instance.

None of this is to say that you aren't having a problem, I'm just not able to visualize how an issue could be cropping up.  I can make two guesses without a reproduction as to the cause, however, I'm just spitballing here.
* There is some duplication of a service's fullname with different GraphQL configurations.  That is, two services exist somewhere in the group of nodes, but they have different GraphQL configurations from each other.  That would cause, depending on sequence of service processing, the potential for the GraphQL schema to be built differently on the new instance of `moleculer-
web` than it did the previous time.
* The second replica of moleculer-web isn't a _true_ replica, but is being launched as a service with a fullname (name or version) that is different from the original, but the GraphQL configurations have overlap with each other.  Therefore, what would happen is both the local services and the remote services would get processed, and effectively duplicate GraphQL types into the schema that is being generated, which would cause a failure.  A GraphQL type must have a single representation or else the schema cannot be built.  If the same GraphQL type is present multiple times, but each time it is present is within a replica of a service with the same name and version (i.e. fullname), then it should be skipped in the code I laid out earlier which would avoid the duplication.

As I said before, this is just guessing.  In our environment we definitely have multiple instances of `moleculer-web`/`moleculer-apollo-server` and they have their own locally defined schema definitions even if there aren't other services running on the same node.  We don't have a problem with schema generation as the remote service's of the same name are skipped during schema generation.

Is there any way you can provide a stripped down repro that shows the problem?  I'd like to help, but I'm not sure how far I can take this without seeing your actual setup.

**Comment by shawnmcknight** [comment]
No activity on this in several months.  Closing for now.

**Comment by abdavid** [comment]
Sorry for the seriously late reply - I have been hammered with project deadlines and life in general (like most of us).

I have some new insight that might shed some light and I am not certain that it can be resolved without me having to rewrite a bunch of resolvers.

So tonight I have been digging deeper -- I have been playing around locally with the issue. What I have done is split up the moleculer-apollo-server gateway so it runs in its own process. After that I started up the project without the gateway so my services with their respective graphql configuration then would register (using TCP transporter while testing).

My finding (which is also logical when I think about it) is that I have a problem with my resolvers. Some of my resolvers are so simple that I did not use the moleculer-apollo-server "norm" of mapping the resolver to a service, I rather wrote the resolver as an inline function e.g:

```javascript
....
id: source => source.customerNumber,
....
```

And that of course it not serializable for the transporter, so when my moleculer-apollo-server parses the services in the registry to identify and configure the schema - resolvers which are inline functions just gets voided. And I also now see why my fix works because when I start the project and everything i running locally the registry actually has the resolver functions.
